docs/design/2022-07-20-session-manager.md

```diff
@@ -0,0 +1,190 @@
+
+When a TiDB instance quits accidentally, the ongoing queries fail immediately and the Session Manager redirects the connections.
+
+![session manager health check](./imgs/session-manager-health-check.png)
+
+### Authentication
+
+When the Session Manager migrates a session, it needs to authenticate with the new TiDB server.
+
+It's unsafe to save user passwords in the Session Manager, so we use a token-based authentication:
+
+1. The administrator places a self-signed certificate on each TiDB server. The certificate path is defined by a global variable. The certificates on all the servers are the same so that a message encrypted by one server can be decrypted by another.
+2. When the Session Manager is going to migrate a session from one TiDB instance to another, it queries the session token. The session token is composed by the username, token expiration time, and a signature. The signature is signed with the private key of the certificate.
+3. The Session Manager then authenticate with the new TiDB server with a new auth-plugin. The session token acts as the password. The new server checks the username, token expiration time, and the signature. The signature should be verified by the public key.
+
+To ensure security, TiDB needs to guarantee that:
+
+- The certificate rotates periodically so that it minimizes the effect of leaking the certificate.
+- The username in the session token must be the same with the that in the handshake packet, so one user cannot log in with the identity of another user.
+- The token expiration time cannot exceed the current time and the token lifetime cannot be too long, so one user cannot forge the token by brute force, or use a valid token for a long time.
+- The signature should be verified so that TiDB can make sure the data is not forged.
+- Secure transport is enforced when querying the session token because the token cannot be leaked.
+
+### Connection State Maintenance
+
+A MySQL connection is stateful. TiDB maintains a session state for each connection, including session variables, transaction states, and prepared statements. If the Session Manager redirects the frontend query from one backend connection to another without restoring the session state in the new connection, an error may occur.
+
+The basic workflow is as follows:
+
+1. When the client queries from the Session Manager, the Session Manager forwards the commands to TiDB and then forwards the query result from TiDB to the client. The session states are only updated by TiDB.
+2. When the Session Manager is going to migrate a session from one TiDB instance to another, it queries the session states from the original TiDB instance and saves them.
+3. The Session Manager then connects to the new TiDB instance and replays the session states.
```

Comment:
Similarly, maybe introduce `show @@session_states` explicitly in the design doc. Otherwise it works like unknown magic.